[로또 2단계] 채넛 미션 제출합니다.

yrsel opened the pull request:
# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

## 코드 리뷰 커뮤니케이션

안녕하세요, 페로로 !
모바일 안드로이드 크루 채넛 입니다 🙇‍♂️

2단계 로또 수동 기능이 추가되면서 고민되는 부분이 생겨 고민하다보니 리뷰 요청까지 오랜 시간이 걸렸습니다. 🫠
1단계 리뷰해주신 부분 반영하고 2단계 수동 기능까지 구현하였습니다.

1단계 리뷰를 다시 읽어보며 키오스크 예시와 예외 처리에 대한 코멘트가 2단계 미션 내용 하단에 작성되어 있던 논리적 오류와 비논리적 오류에 
대한 이해 그리고 `sealed class` 활용한 예외처리에 유연성을 더하는 방향을 의미하셨던 것 같은데 맞을까요 ,,? 
아직 어떤 방식으로 예외처리하면 좋을지에 대한 고민은 진행중이여서 학습하며 다음 리뷰에서 적용해보도록 하겠습니다.
그리고, 2단계를 진행할 때 고민되었던 사항에 대해 질문 드리고 싶어 아래 글을 남겼습니다 !

# 어떤 부분에 집중하여 리뷰해야 할까요?

로또 생성기의 추상화에 대한 고민이 있었습니다.

1단계에서 `LottoMachine` 인터페이스와 `RandomLottoMachine` 클래스가 존재했었기에 동일하게 로또를 발행하는 역할을 수행할 수동 로또 생성기도 `LottoMachine` 인터페이스의 구현체로 설계할 계획이었습니다.

하지만 구현 메서드로 `createLotto(), createLottos(count : Int)` 와 같이 존재했습니다.
로또 하나를 생성하는 `createLotto()` 를 구현하려는 과정에서 차이점을 설명하자면
랜덤로또 생성기: 로또 생성시에 내부 랜덤 함수를 실행하기에 매개변수가 따로 필요 없다.
수동로또 생성기: 로또 생성시에 사용자가 입력한 값으로 로또로 발행해줘야 하기에 인자로 넘어 올 값이 필요하다.

여기서 인터페이스에서 제공하는 함수 시그니처를 깨지 않기 위해서는 `수동 로또 생성기`의 생성자로 사용자의 입력을 받는 방법이 있을 것 같습니다.
→ 하지만, 이 방법의 문제는 사용자가 잘못된 로또 번호를 입력하게 되어도 **모든 입력을 받은 후에야 알 수 있다는 점** + **다시 모든 수동 로또 번호를 입력해야 한다는 점**이 있습니다.
→ 그렇다면 하나씩 검증하면 되지 않는가..?  
→ `수동 로또 생성기` 생성자로 사용자 입력을 하나씩 받으며 로또를 생성한다면 매번 `수동 로또 생성기 인스턴스`를 생성하며 로또를 발행해야 한다는 단점이 존재합니다. (매번 수동생성기 인스턴스 생성해야 하는 부분이 이상한 것 같습니다.)  
→ 이러한 생각의 흐름의 결과로 인터페이스 추상화를 포기하고 `AutomaticLottoMachine` 과 `ManualLottoMachine` 각각의 클래스를 생성해서 로또를 발행하도록 구현하였습니다.

- 로또를 발행한다는 역할이 동일하고 내부 구현만 달라지는 것 같아 추상화를 진행하면 좋을 것 같다는 생각이었는데 불필요한 집착이었던 것 일까요 ?
- 페로로의 경우 추상화하는 기준을 어떤 기준으로 잡으시는 지 궁금하고 현재 저의 생각에 대한 의견도 부탁드립니다..! : )

로또 2단계에서도 많이 얻어가겠습니다, 피드백 감사합니다 ! : )

## 스크린샷
- 전체 실행 결과
![스크린샷 2025-02-27 00 11 42](https://github.com/user-attachments/assets/6b9b7077-8751-4ecc-b57d-16d832bfc861)

yrsel committed:
refactor: 로또 데이터 구성에 List 빌더 함수 활용

yrsel committed:
refactor: 로또 번호 value 클래스로 변경

yrsel committed:
refactor: 구입 금액을 당첨 통계 계산 할 때 넘겨주도록 수정

yrsel committed:
refactor: 통계 출력 부분 MISS 제외 로직 수정

yrsel committed:
test: 사용되지 않는 setUp 함수 제거

yrsel committed:
refactor: 총 상금 계산하는 로직 수정

yrsel committed:
refactor: 적절하지 못한 캐싱 위치의 코드 제거

yrsel committed:
refactor: 요구사항 변경에 따른 코드 제거

금액을 로또 하나의 금액으로 나눈 나머지가 0인지 검증하는 로직 제거

yrsel committed:
docs: 로또 구매 수량을 담당하는 기능 추가

yrsel committed:
feat: 구매 가능한 로또 수량에 대한 판단 구현

yrsel committed:
feat: 로또 수량 차감에 대한 로직 구현

yrsel committed:
feat: 구매 수량 생성자 검증 로직 구현

yrsel committed:
feat: 수동 로또 입력 구현

yrsel committed:
refactor: LottoCount 데이터 클래스로 변경

yrsel committed:
refactor: 매개변수 LottoCount 일급객체로 변경

yrsel committed:
refactor: 예외를 LottoCount 내부에서만 검증

yrsel committed:
refactor: 중복되는 숫자 변환 로직을 함수로 치환

yrsel committed:
refactor: 중복되는 로또 숫자 묶음 변환 로직을 함수로 치환

yrsel committed:
feat: 로또 번호 캐싱 및 정적 팩토리 메서드 적용

yrsel committed:
remove: 로또 머신 인터페이스 제거

yrsel committed:
rename: RandomLottoMachine.kt -> AutomaticLottoMachine.kt으로 변경

yrsel committed:
refactor: 로또 묶음을 발행해주는 함수 변경 및 싱글턴 객체로 변경

yrsel committed:
feat: 수동 로또 발행 역할 구현

yrsel committed:
rename: readManualLotto() -> readManualLottoNumbers()

yrsel committed:
fix: 컨트롤러 생성자 매개변수 수정

yrsel committed:
rename: getLottoCount() -> getWholeLottoCount()

yrsel committed:
feat: 수동 로또 구매 안내 문구 출력 구현

yrsel committed:
feat: 수동 로또 구매 로직 구현

yrsel committed:
feat: 자동 로또 구매 로직 구현

yrsel committed:
feat: 로또 출력 관련 로직 구현

yrsel committed:
feat: 로또 숫자가 양수 판단 여부 구현

yrsel committed:
chore: 불필요한 import 제거

yrsel committed:
feat: 전체 로또 반환 함수 구현

yrsel committed:
feat: 소수점 버림 로직 구현

yrsel committed:
refactor: 수동 로또 생성시 정렬 로직 구현

yrsel committed:
style: 가독성 위한 선언 위치 수정

yrsel committed:
test: 수익률 로직 수정에 따라 테스트 코드 예상값 수정

laco-dev commented on README.md:
```diff
@@ -26,10 +26,13 @@
     - [x] 구입 금액으로 구매할 수 있는 로또 개수를 반환한다.
+- 로또 구매 수량
+    - [x] 로또 구매 수량은 0보다 크거나 같다.
+    - [x] 구매할 수 있는 수량인지 판단한다.
```
> 1단계 리뷰를 다시 읽어보며 키오스크 예시와 예외 처리에 대한 코멘트가 2단계 미션 내용 하단에 작성되어 있던 논리적 오류와 비논리적 오류에
대한 이해 그리고 sealed class 활용한 예외처리에 유연성을 더하는 방향을 의미하셨던 것 같은데 맞을까요 ,,?

어떤 내용을 이야기 하고 싶은지 정확히 기억이 안나기에, 리뷰 링크를 같이 남겨주시면 좋겠어요.
키오스크라면 InputView, OutputView의 역할이 모호한 것으로 아마 의견을 드렸을 것 같은데요.
논리적인 오류와는 관계 없는 부분입니다.

yrsel replied:
https://github.com/woowacourse/kotlin-lotto/pull/108#discussion_r1966622956

자주 읽다보니 익숙해져서 상대방을 고려하지 못하고 두서없이 작성했네요. 😅

남겨주신 리뷰에서 키오스크 예시와 더불어 `단순히 IllegalArgumentException으로 재고가 부족하다는 것을 퉁치는것이 문제가 되었던 것 같습니다.` 라는 리뷰를 작성해주셨는데 비논리적인 오류(사용자의 입력으로 인해 발생한 오류)에 대한 처리를 조금 더 고민해보라는 의미로 착각했습니다.

laco-dev commented on src/main/kotlin/lotto/controller/LottoController.kt:
```diff
@@ -36,14 +36,55 @@ class LottoController(
+        try {
+            outputView.printPurchaseManualLottoCountGuide()
+            val manualLottoCount = inputView.readManualLottoPurchaseCount()
+            wholeLottoCount.validateLottoCount(manualLottoCount)
```
LottoCount의 역할은 Validator가 아닙니다.
우리가 알고 싶은 기능은 로또를 몇 장 살 수 있는지가 아닐까요?

그렇다면, 예외를 발생시키는 대신 구입 가능한지 여부를 판단하는 기능을 만들어보는게 좋을 것 같아요

> test : 로또 갯수가 입력한 갯수보다 적은지 알 수 있다
```kotlin
val manualLottoCount = inputView.readManualLottoPurchaseCount()
if (wholeLottCount.value < manualLottoCount) {
   // Retry
}
```

yrsel replied:
객체 내부에 구현을 숨기는 것이 좋다고 생각해 예상 결과가 아닐 경우 예외를 발생시켜서 종료하는 방식을 주로 택했던 것 같습니다.
리뷰 방식대로 진행하는 것이 좋을 것 같아서 `LottoCount` 객체에게 구매 가능여부를 묻고 구매가 어렵다면 입력을 다시 받도록 수정하였습니다 ! 

관련 커밋 : [b88657](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/b8865737b0ee7d0ec5e8fa9a01d47526da84932f)

laco-dev commented on src/main/kotlin/lotto/controller/LottoController.kt:
```diff
@@ -36,14 +36,55 @@ class LottoController(
+            getPurchaseManualLottoCount(wholeLottoCount)
+        }
+
+    private fun getManualLottoBundle(manualLottoCount: LottoCount): List<Lotto> {
+        if (manualLottoCount.isPositive()) outputView.printManualLottoNumbersGuide()
+        return List(manualLottoCount.number) { getManualLotto() }
+    }
```
LottoCount가 논리적으로 항상 0보다 크도록 작성하였어요. 그렇다면 isPositive() 라는 개념이 필요할까요?

yrsel replied:
관련 커밋 : [183857](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/183857c2fb68c4f19aeb2286bf2fd179bb3de1a1)

laco-dev commented on src/main/kotlin/lotto/model/LottoCount.kt:
```diff
@@ -0,0 +1,24 @@
+        require(isAvailPurchaseLottoCount(lottoCount)) { ERROR_EXCEED_LOTTO_COUNT }
+    }
+
+    fun minus(lottoCount: LottoCount): LottoCount = LottoCount(this.number - lottoCount.number)
```
예외를 발생시키지 않고 논리적인 개념으로 접근해본다면 이런 예시를 작성해볼 수 있습니다.
숫자는 0보다 작지 않다. 라는 논리적인 흐름을 유지하는 것이죠

> Test : 로또 숫자 5에서 로또 숫자 10을 빼더라도 로또 숫자는 0보다 작을 수 없다.

yrsel replied:
관련 커밋 
-  [5dbd1c](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/5dbd1c6961d84a9f0ac01af3a120a789af8dbbac)
-  [3416f7](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/3416f7c2eea361405b3dd210f18a64f93d48646d)

laco-dev commented on src/main/kotlin/lotto/model/lottomachine/AutomaticLottoMachine.kt:
```diff
@@ -0,0 +1,21 @@
+import lotto.model.LottoNumber.Companion.MAX_NUMBER
+import lotto.model.LottoNumber.Companion.MIN_NUMBER
+
+object AutomaticLottoMachine {
```
> 수동 로또 생성기 생성자로 사용자 입력을 하나씩 받으며 로또를 생성한다면 매번 수동 로또 생성기 인스턴스를 생성하며 로또를 발행해야 한다는 단점이 존재합니다. (매번 수동생성기 인스턴스 생성해야 하는 부분이 이상한 것 같습니다.)
로또를 발행한다는 역할이 동일하고 내부 구현만 달라지는 것 같아 추상화를 진행하면 좋을 것 같다는 생각이었는데 불필요한 집착이었던 것 일까요 ?

생성자로 받으면 매번 인스턴스를 생성해야 한다는 단점을 잘 캐치하셨습니다.
로또 머신은 로또를 발행한다. 라는 역할을 구현하고 싶은데 고민하고 계신 여러 문제점에 직면하게 됩니다.

아직은 협력 관계에 대해 익숙하지 않으시기 때문에 발생할 수 있다고 생각합니다.
객체지향적으로 더 깊이있게 생각한다면 이런 구조도 만들 수 있다는 예시만 참고로 보여드립니다.
이 구조를 따라 만드려고 하지 않으셔도 괜찮아요
점차 학습을 하시면서 시야를 넓혀 가시면 되겠습니다.

```kotlin

interface LottoMachine {
   fun create(size: Int): List<Lotto>
}

interface LottoNumberProvider {
   fun get(size: Int): List<LottoNumber>
}

class ManualLottoMachine(
  private val provider: LottoNumberProvider
): LottoMachine {
   override fun create(size: Int): List<Lotto> {
      ...
      val lottoNumbers = provider.get(6)
      val lotto = Lotto(lottoNumbers)
   }
}
```

laco-dev commented on src/main/kotlin/lotto/view/InputView.kt:
```diff
@@ -1,29 +1,41 @@
 class InputView {
     fun readLottoPurchaseAmount(): LottoPurchaseAmount {
-        val money = readln().trim().toIntOrNull() ?: throw IllegalArgumentException(ERROR_NOT_A_NUMBER)
+        val money = readln().toIntOrException()
```
모든 것을 Exception으로 처리하려는 습관이 잘못 들까봐 의견을 드려봅니다.

int로 변환할 수 없으면, 예외를 터트려버릴 때 이것을 사용하는곳에서 할 수 있는 액션이 있을까요?
이를 논리적인 오류라고 인지할 것인지, 아니면 충분히 무시할 수 있는 상황이므로 개발의 유연성을 지킬 것인지에 따라 다를 수 있습니다.

만약 문자열을 입력해서 변환할 수 없었다면, 이는 InputView에서 스스로 판단하여 다시 재입력 받도록 할 수 있을 것이지만,
만약 LottoPurchaseAmount를 생성하면서 문제가 발생한다면 이는 경우에 따라 달라질 것 같아요.

아마 설명을 들으셔도 지금은 잘 이해가 안될거라고 생각합니다.
논리/비논리적인 문제로 너무 고민 하지는 않으셔도 되구요. 앞으로 경험할 기회가 많으니 나중에 자신의 코드를 스스로 돌아보는 정도면 충분히 학습이 될 것 같습니다.

yrsel replied:
예외로 모든 문제를 해결하려는 습관이 있었던 것 같습니다.
사용자 입력에서 발생한 문제는 비논리적 오류라고 판단해 엘비스 연산에서 재귀를 진행해 입력을 다시 받을 수 있도록 변경하였고 로또 생성과 같이 프로그래머가 제한할 수 있는 환경에서 발생한 에러의 경우 논리적 오류라고 판단해 runCatching 메서드를 통해 발생한 예외를 처리한 후 재귀를 진행하도록 하였습니다.

말씀해주신 내용에 대해 아직 논리, 비논리 오류에 대한 기준이 명확하게 잡히지 않은 것 같습니다.
나중에 깨달음을 얻고 돌아오겠습니다 ! 

관련 커밋 : [532015](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/5320152843658e52a93db7b6df35e06794fdc83a)

laco-dev commented on src/test/kotlin/lotto/LottoCountTest.kt:
```diff
@@ -0,0 +1,33 @@
+class LottoCountTest {
+    @Test
+    fun `로또 수량은 0보다 크거나 같아야 한다`() {
+        assertThrows<java.lang.IllegalArgumentException> { LottoCount(-1) }
```
PR을 올리고 나더라도, 자기가 작성한 코드 한번씩 정리하면서 살펴보세요.
이런 부분은 스스로 잡으셔야 합니다.

yrsel replied:
관련 커밋 : [32f369](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/32f3698f11dbec4d14ad2696793b4a85627a7298)

laco-dev reviewed (changes requested):
2단계 고생 많으셨습니다.

객체지향적인 사고나 논리적인 문제에 대하 고민하는 모습이 인상깊었습니다.
전반적으로 작성은 잘 해주셨지만 모든 것을 오류로 해결하려는 방법은 조금 아쉬운 것 같아요.

질문주신 내용과 같이 의견을 남겨두었습니다.

yrsel committed:
remove: 로또 수량 검증하는 함수 제거

yrsel committed:
refactor: 구매 가능한 로또 수량인지 판단 로직 구현

yrsel committed:
remove: 로또 수량이 0보다 큰 수인지 판단 로직 제거

yrsel committed:
refactor: 로또 수량 논리적 개념 정립

yrsel committed:
refactor: 초기화 방식 수정

커스텀 getter로 외부 노출 되는 데이터에 대한 값만 0으로 변경해주니 음수 인스턴스가 생성될 경우 외부 노출되는 number 값은 0으로 동일하나 생성자로 받는 값이 음수가 되어서 인스턴스 일치여부를 판단할 수 없었다. -> 초기화 블럭에서 0으로 초기화 진행하도록 변경

yrsel committed:
fix: 패키지명 제거

yrsel committed:
refactor: 예외 상황에 대한 구현 변경

yrsel commented:
페로로 !
질문이 있습니다 🙋‍♂️

현재 코드 작성 방식으로 진행하게되면 모든 문제를 예외로 처리하는 습관이 생길 수 있다는 의견에 몹시 공감했습니다.
그렇기에 상황에 따라 예외가 아닌 다른 방법으로 처리하는 방식을 고민해보았는데요.
우선 제가 생각하는 논리적 오류와 비논리적 오류 측면으로 처리방식을 나눠보았습니다.

논리적 오류는 현재 프로그램에서 의도한 방향대로 흘러가지 않고 프로그램이 작동하는 경우라고 생각했습니다.
예를 들어, `로또 번호는 1~45 사이의 값이어야 한다.` 라는 규칙이 있다면 -1, 0, 46 등 규칙에서 벗어난 로또 번호가 생성되어 프로그램이 동작은 하지만 의도한 방향은 아니기에 논리적인 오류다 라고 정의하였고 이럴 경우에는 예외를 반환하고 예외가 발생한 원인에 대해 설명하고나서 다시 입력을 받도록 하였습니다.

비논리적 오류는 프로그래머가 제어할 수 없는 데이터에 의해 생긴 오류라고 생각했습니다.
예를 들자면, `프로그램에서는 숫자를 기대했는데 사용자에 의해서 문자가 들어온다.` 이러한 경우 비논리적 오류라고 생각했고 예외가 아닌 재귀를 통해 입력을 다시 받도록 하였습니다.

제가 생각한 논리적 오류, 비논리적 오류가 발생했을 때 처리 방식이 괜찮은 지 어떤지 궁금합니다.
생각의 도움이 될 수 있는 자그마한 힌트라도 주시면 감사하겠습니다 ! :-)

그 외 다른 코멘트 필요한 부분도 잘 부탁드리겠습니다. 감사합니다 🙇‍♂️

laco-dev commented on src/main/kotlin/lotto/model/LottoCount.kt:
```diff
@@ -0,0 +1,15 @@
+package lotto.model
+
+data class LottoCount(
+    private var _number: Int,
+) {
+    val number = _number
+
+    init {
+        if (_number < 0) _number = 0
+    }
```
> 객체 내부에 구현을 숨기는 것이 좋다고 생각해 예상 결과가 아닐 경우 예외를 발생시켜서 종료하는 방식을 주로 택했던 것 같습니다.
리뷰 방식대로 진행하는 것이 좋을 것 같아서 LottoCount 객체에게 구매 가능여부를 묻고 구매가 어렵다면 입력을 다시 받도록 수정하였습니다 !

논리적으로 LottoCount가 0이 될 수 없다는 점을 지키기 위해 불변 클래스로 만드는 연습이 필요할 것 같네요!
data class는 기본적으로 ReadOnly를 지향합니다.
그렇기 때문에 내부적으로 상태를 변경시키는 방식이 아닌. 적절한 구현을 통해 이를 유지하게 됩니다.

```kotlin
data class LottoCount(val number: Int) {
   init { require(number >= 0) }

   fun minus(other: LottoCount): LottoCount {
      val number = (this.number - other.number).coerceAtLeast(0)
      return LottoCount(number)
   }
}
```

yrsel replied:
https://github.com/woowacourse/kotlin-lotto/pull/143#discussion_r1972830262

해당 코멘트에서 `예외를 발생시키지 않는다`에 홀렸나봐요.
require 를 통해 예외처리 하지 말라는 의미로 잘못 해석하는 바람에 초기화 블럭에서 강제로 0으로 변경해주었습니다.

minus() 메서드에서 논리적으로 0이 될 수 없음을 보장하였고 객체 생성자에 음수로 입력하는 경우를 제한하기 위해 초기화 블럭에서 예외처리 진행하도록 수정하였습니다 ! 

관련 커밋 : [446e8a](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/446e8adbc5a3aff6598dc2458aba85bb698cb62d)

laco-dev commented on src/main/kotlin/lotto/model/LottoNumber.kt:
```diff
@@ -11,5 +12,9 @@ data class LottoNumber(
+
+        private val cachedLottoNumber = (MIN_NUMBER..MAX_NUMBER).associateWith { number -> LottoNumber(number) }
+
+        fun from(number: Int): LottoNumber = cachedLottoNumber[number] ?: throw IllegalArgumentException(ERROR_OUT_OF_RANGE_NUMBERS)
```
from, of, valueOf등 팩토리 생성을 하기 위해 여러가지 이름이 있는데 찾아보셨엇나요?

yrsel replied:
넵, 정적 팩토리 메소드 패턴에 대해 학습할 때 명명규칙에 대한 내용을 본 적 있습니다 !
`from` : 하나의 매개변수를 받아 객체를 생성
`of` : 여러 매개변수를 받아 객체를 생성
`valueOf` : 객체 생성에 조금 더 의미를 부여하는 이름 규칙 
이라고 알고 있습니다.

from, of 에 대한 이름 규칙을 모르는 경우도 있으니 현재 상황에서 조금 더 직관적인 valueOf 라는 메서드명이 더 좋았을까요 ??

laco-dev replied:
아니요. 특징을 이해하고있는지 묻는 질문이었습니다. :)

laco-dev commented on src/main/kotlin/lotto/model/LottoStatistics.kt:
```diff
@@ -9,22 +12,19 @@ class LottoStatistics(
-        }
+    fun isLossMoney(): Boolean = rateOfReturn < 1
+
+    private fun getTotalPrize(): Double = rankStatistics.asIterable().sumOf { (rank, count) -> rank.prizeMoney * count }.toDouble()
```
함수형 프로그래밍이 코드 작성에 편리한 부분을 제공해 주는 점은 맞지만. 한편으로는 가독성 관점에서 비교하면 아쉬운점이 많을 때도 있어요.

기존 코드와 비교했을 때 지금 코드는 어떤점이 좋아졌다고 생각하시는지 궁금합니다!

yrsel replied:
맞습니다, 저도 함수형 프로그래밍을 처음 접했을 때 어떻게 동작하는 건지 감도 오질 않고 오히려 코드 읽기가 어려웠던 기억이 있습니다. 🥲
그렇지만 함수형 프로그래밍에 조금씩 익숙해지면서 동일한 내용을 함축적으로 짧은 코드로 표현할 수 있는 점이 매력적으로 느껴지는 것 같습니다.

기존 코드와 비교했을 때 합계를 담을 **변수명을 고려할 필요가 없다**는 점, 다양한 메서드 체이닝 연산들을 제공한다는 점, 제공되는 메서드명이 행동을 잘 표현하고 있다는 점이 장점이라고 생각합니다.

`asIterable()` 메서드 대신 Map 자료구조를 활용하는 방향인 `Map.Entry<Key,Value>` 값을 활용하기 위해 `entries` 로 수정하였습니다. 
이렇게 적고보니 과연 모두가 이해하기 좋은 코드인가 싶은 생각이 들기는 하네요.😭

관련 커밋 : [d4443f](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/d4443f333773d31c6047fc51f906ea14d111e937)

laco-dev commented on src/main/kotlin/lotto/view/InputView.kt:
```diff
@@ -1,31 +1,73 @@
 import lotto.model.LottoPurchaseAmount
 
 class InputView {
+    private val outputView = OutputView()
+
```
콘솔에출력하는 것을 반드시 OutputView만 해야하는 것인가요?
InputView는 println에 접근하면 안되는 것인지 의견이 궁금합니다.

yrsel replied:
이 부분도 고민했던 부분입니다.
제거해도 좋을 것 같다고 생각해서 `InputView`에서 println으로 출력을 진행했었습니다.
`InputView`에서도 예외 문구를 출력하는 부분이 존재하게 되었고 `OutputView`에서 사용하는 예외 출력 메서드를 동일하게 사용하고 싶어서 `OutputView`객체를 가져와서 사용하게되었습니다.
`OutputView`를 참조하게 되면서 모두 `OutputView`의 메서드를 호출할 수 있도록 변경하였습니다.

추가적으로 같은 레벨에 있다고 생각되며 입력에서 출력에 대해 알고 있는 정도는 괜찮다고 생각했습니다 !

laco-dev replied:
네 좋습니다.

지금은 InputView와 OutputView가 역할을 명확하게 나누려고 하다보니 어색하지만 실제로는 구분을 하지 않고 하나의 개념으로 사용이 되고는 합니다.

laco-dev commented on src/main/kotlin/lotto/view/InputView.kt:
```diff
@@ -1,31 +1,73 @@
+    }
+
+    private fun readLottoNumbersWithGuideMessage(guideMessage: () -> Unit): List<LottoNumber> {
+        guideMessage()
```
guideMessage()가 뭘 하고싶은건지 의도가 잘 드러나는 이름인가요?
저는 상태처럼 느껴지는데요! 앞서 OutputView때문에 이런 구조가 된 것 같은데 조금 아쉽군요!

yrsel replied:
의도가 잘 드러나지 않아서 수정하였습니다 !
`OutputView`의 파급효과네요...

관련 커밋 : [75f35c](https://github.com/woowacourse/kotlin-lotto/pull/143/commits/75f35cc05f6ed0e36a66b172a52ef0dad3d925ce)

laco-dev commented on README.md:
```diff
@@ -26,10 +26,13 @@
+- 로또 구매 수량
+    - [x] 로또 구매 수량은 0보다 크거나 같다.
+    - [x] 구매할 수 있는 수량인지 판단한다.
+    - [x] 현재 수량에서 구매한 수량을 차감한다.
```
> 논리적 오류는 현재 프로그램에서 의도한 방향대로 흘러가지 않고 프로그램이 작동하는 경우라고 생각했습니다.
예를 들어, 로또 번호는 1~45 사이의 값이어야 한다. 라는 규칙이 있다면 -1, 0, 46 등 규칙에서 벗어난 로또 번호가 생성되어 프로그램이 동작은 하지만 의도한 방향은 아니기에 논리적인 오류다 라고 정의하였고 이럴 경우에는 예외를 반환하고 예외가 발생한 원인에 대해 설명하고나서 다시 입력을 받도록 하였습니다.

> 비논리적 오류는 프로그래머가 제어할 수 없는 데이터에 의해 생긴 오류라고 생각했습니다.
예를 들자면, 프로그램에서는 숫자를 기대했는데 사용자에 의해서 문자가 들어온다. 이러한 경우 비논리적 오류라고 생각했고 예외가 아닌 재귀를 통해 입력을 다시 받도록 하였습니다.

좋아요! 잘 따라오고 있습니다.
쉬운 힌트를 드려보면 테스트를 작성해야 한다는 생각이 마구마구 솟는다면 == 논리적인 오류 입니다.
그렇지 않고 코드를 작성하다가 누가봐도 문제는 없어보인다면 no-op을 하거나 코드가 자연스럽게 이어지도록 적당히 처리를 하고는 합니다.

프로그램이 서비스중이라면, 논리적인 오류가 발생하는 상황은 반드시 인지해야 하는 중요한 상황일거에요.
오류 리포팅을 하거나 해서 수정하지 않으면 고객센터 문의가 빗발치겠죠?

가령 서버에서 남은 잔액을 보내주는데 잔액이 마이너스다?
그게 UI에 자연스럽게 마이너스로 표시되면 고객은 아마 화를 낼겁니다 😠

yrsel replied:
예시를 보니 조금 더 명확하게 이해 되는 것 같습니다 ! 
감사합니다 ! :-)

laco-dev reviewed (changes requested):
피드백 반영 잘 해주셨어요.
스스로 고민하고 깨닫는게 엄청 빠르시네요!

이후 미션들에서도 조금씩 더 배워나가면 되니 너무 고민하지 말고 지금은 컨디션 조절 하시는것을 추천드려요!

어느정도 구조가 잡힌 것 같아서 코드 정리만 하면 미션 마무리 할 수 있을 것 같습니다.

yrsel committed:
refactor: 로또 수량 개념 정리

로또 수량은 0보다 작을 수 없다.

yrsel committed:
fix: 총 당첨 금액 메서드 내부 구현 수정

yrsel committed:
rename: 가이드 메시지 출력 역할의 파라미터명 수정

laco-dev reviewed (approved):
고생 많으셨습니다 :)
지금까지 배운 내용들은 이후 미션에서도 잘 지킬 수 있도록 기억해주셨으면 좋겠어요.
다음 미션들도 응원하겠습니다 👍🏻
